### ISSUE TITLE
How to handle boundary conditions nicely? User-friendly API + kernels that support multiple BCs.

### ISSUE THREAD

**Issue by ali-ramadhan** [climate-machine/Oceananigans.jl#86]
Would be nice to have a nice and user-friendly API for setting boundary conditions. What boundary conditions should we be supporting? What should this API look like? Just starting a discussion around this that I've had with @SandreOuza and @glwagner.

@glwagner made the point that the user should just have to specify whether they'd like Dirchlet, Neumann, or Robin boundary conditions. Periodic should be another option. We don't have to support all combinations.

But in finite volume boundary conditions are converted into conditions on the fluxes.

Right now we have something like this
```julia
model.boundary_conditions = BoundaryConditions(x=:periodic, y=:periodic, top=:free_slip, bottom=:free_slip)
```
and all the ugliness is swept under the rug.

This may involve refactoring some of the difference and interpolation operators as they might have been implicitly assuming that free-slip is being imposed. Will have to confirm with @jm-c.

PS: Assigning multiple people as I don't have much experience around this (just to get your attention) Would be good to talk and figure out what kind of API works for all of us.

Pinging: @SandreOuza

**Comment by glwagner** [comment]
I propose that we develop a type that corresponds to a boundary conditions. I think the following user code will be nice:

```Julia
...
surface_flux(x, y, z, t) = F_b * exp(-x^2 / 2L^2) * cos(σ*t)
bottom_flux = N^2

top_bc = DirichletBoundaryCondition(:T, :z; :right, surface_flux)
bottom_bc = DirichletBoundaryCondition(:T, :z, :left, bottom_flux)

model = Model(... bcs=[top_bc, bottom_bc])
```

(The names of the types are certainly up for debate: we might also have `ConstantDirichletBoundaryCondition` and `VariableDirichletBoundaryCondition`, and either expose it to the user or hide this layer from from the user and use multiple dispatch to parse between user-inputs.)

We can propose default boundary conditions that are deployed whenever a boundary condition is not specified.

**Comment by ali-ramadhan** [comment]
That's a beautiful API! Users shouldn't have to worry about how to impose boundary conditions in a finite volume (or discontinuous Galerkin) formulation. The boundary conditions should be specified as you would specify them mathematically (I'm just repeating what you're saying).

Instead of `:right` and `:left` did you mean `:top` and `:bottom`?

+1 to using default boundary conditions when they're not specified. Could be mentioned in the docs and log #71 to alert the user that this was done.

**Comment by glwagner** [comment]
I did not mean top or bottom — every coordinate comes with a sense of direction, so the words “left” and “right” are generally applicable to every coordinate system. There is no need for a special syntax for certain directions.

**Comment by ali-ramadhan** [comment]
Ah maybe I'm misunderstanding but why does a boundary condition on the surface `top_bc = DirichletBoundaryCondition(:T, :z; :right, surface_flux)` need to know something about `:left` or `:right`?

**Comment by glwagner** [comment]
A “boundary condition” is a condition that applies at one of the ends of a certain coordinate. So a boundary condition can be applied at the “right” (positive end) or “left” (negative end) of Cartesian x, y, and z, latitude, longitude, buoyancy coordinate etc.

**Comment by ali-ramadhan** [comment]
Ah I see your point now. I still feel that using `:left` and `:right` for the _x_ and _y_ directions and `:top` and `:bottom` for the _z_-direction is clearer but this is a minor point.

**Comment by glwagner** [comment]
We can design a functionality that just uses left and right (which is
simple and general) and then write an additional layer over that to allow
users to use top and bottom in their scripts should they feel that is
better.

**Comment by ali-ramadhan** [comment]
@glwagner should this be closed now that #118 has been merged?

**Comment by glwagner** [comment]
Yes.
